elastic/elastic-agent
# Reduce the amount of stored ReplicaSet data

## What does this PR do?

Use a transform function to drop all ReplicaSet data in the kubernetes provider, except for the owner references, which we need to find the Deployment name.

Unfortunately, the autodiscovery library doesn't let us pass the transform function in, so I had to copy their implementation and make the change myself. The plan is to upstream this change later.

## Why is it important?

In clusters with a lot of Deployments, we end up storing a lot of ReplicaSets in the local cache of each agent, resulting in significant unnecessary memory consumption.

## Checklist




## Disruptive User Impact



## How to test this PR locally

There isn't an easy way. You can start a local K8s cluster, create a lot of Deployments, deploy Agent using the Helm Chart [here](https://github.com/elastic/elastic-agent-service/tree/main/helm), and check the memory consumption.

## Related issues


- Partially resolves #4729 
- Closes https://github.com/elastic/elastic-agent/issues/5623
